WcaOnRails/app/models/competition_event.rb

```diff
@@ -3,4 +3,19 @@ class CompetitionEvent < ApplicationRecord
   belongs_to :competition
   belongs_to :event
   has_many :registration_competition_events, dependent: :destroy
+  has_many :rounds, -> { order(:nth_round) }
+  accepts_nested_attributes_for :rounds, allow_destroy: true
+
+  validate do
+    remaining_rounds = rounds.reject(&:marked_for_destruction?)
```

Comment:
When deleting nested attributes, ActiveRecord first marks them for destruction, then runs your validations, and then actually does the database work. This means your validations need to be aware that the stuff it's validating might be able to be deleted. I think this is terribly confusing and means you have to remember to treat associations in a special way, but AFAIK, that's just the way this ORM (maybe all ORMs?) work.

This article explains how things work, but not why: http://homeonrails.com/2012/10/validating-nested-associations-in-rails/.